anonlink/blocking.py

```diff
@@ -0,0 +1,238 @@
+from itertools import chain, product, repeat
+from numbers import Real
+from random import Random
+from typing import (Any, Callable, Hashable, Iterable, List,
+                    Optional, overload, Sequence, Tuple, TypeVar, Union)
+
+
+BlockingFunction = Callable[[int, int, Sequence[bool]], Iterable[Hashable]]
+
+
+_T = TypeVar('_T')
+def _evalf(__funcs : Iterable[Callable[..., _T]],
+           *args,
+           **kwargs) -> Iterable[_T]:
+    return (f(*args, **kwargs) for f in __funcs)
+
+
+class block_and:
+    """ Conjunction of multiple blocking functions.
+
+        Records share a block if they share a block in all of the
+        functions.
+
+        :param funcs: Functions whose conjunction we return.
+
+        :return: The blocking function.
+    """
+    __slots__ = '_funcs'
+
+    @overload
+    def __init__(self,
+                 funcs: Iterable[BlockingFunction]
+                 ) -> None:
+        pass
+
+    @overload
+    def __init__(self,
+                 func: BlockingFunction,
+                 *funcs: Iterable[BlockingFunction]
+                 ) -> None:
+        pass
+
+    def __init__(self, *args):
+        if len(args) == 0:
+            raise ValueError('at least one argument required')
+        elif len(args) == 1:
+            self._funcs = tuple(args[0])
+        else:
+            self._funcs = args
+
+    def __call__(self,
+                 dataset_index: int,
+                 record_index: int,
+                 hash_: Sequence[bool]
+                 ) -> Iterable[Hashable]:
+        funcs_eval = _evalf(self._funcs, dataset_index, record_index, hash_)
+        funcs_enum = (zip(repeat(i), f) for i, f in enumerate(funcs_eval))
+        return product(*funcs_enum)
+
+
+class block_or:
+    """ Disjunction of multiple blocking functions.
+
+        Records share a block if they share a block in any of the
+        functions.
+
+        :param funcs: Functions whose disjunction we return.
+
+        :return: The blocking function.
+    """
+    __slots__ = '_funcs'
+
+    @overload
+    def __init__(self,
+                 funcs: Iterable[BlockingFunction]
+                 ) -> None:
+        pass
+
+    @overload
+    def __init__(self,
+                 func: BlockingFunction,
+                 *funcs: Iterable[BlockingFunction]
+                 ) -> None:
+        pass
+
+    def __init__(self, *args):
+        if len(args) == 0:
+            raise ValueError('at least one argument required')
+        elif len(args) == 1:
+            self._funcs = tuple(args[0])
+        else:
+            self._funcs = args
```

Comment:
The fancy cool overlead type signatures and `__init__` method are entirely the same in these - so this should be shared code.

Reply:
Done with an abstract base class.